File: server/hermesBridge.ts
import { execFile, execFileSync, type ChildProcess } from "child_process";
import fs from "fs";
import net from "net";
import path from "path";

/**
 * Hermes Bridge — the connection layer that lets JARVIS delegate work to Hermes.
 *
 * Hermes (NousResearch/Hermes-Function-Calling) is an autonomous agent CLI
 * capable of multi-step tool calling, web search, terminal execution, and
 * code writing. This bridge allows JARVIS to dispatch complex, multi-turn
 * reasoning or research tasks to Hermes as a sub-agent.
 *
 * Two communication paths:
 *   1. CLI delegation (primary): `hermes chat --query-file` — spawns Hermes headless
 *      for a single delegated task and returns the final text response. Robust,
 *      always available, no extra services required.
 *   2. Live gateway (optional): a running `hermes serve` JSON-RPC/WebSocket gateway
 *      on 127.0.0.1:9119. When present we report it as "connected" so the UI can show
 *      Hermes as online. The CLI path still executes the actual work; the gateway is
 *      the liveness/status signal.
 */

const HERMES_BIN = process.env.HERMES_BIN || "hermes";
const TIMEOUT_MS = Number(process.env.HERMES_TIMEOUT_MS) || 180_000;
const MAX_TURNS = Number(process.env.HERMES_MAX_TURNS) || 12;
// Default gateway URL for the single Hermes profile
const GATEWAY_URL = process.env.HERMES_GATEWAY_URL || "127.0.0.1:9119";
const OMH_TOOLSET_WARN = /Warning:\s*Unknown toolsets:\s*omh/i;

export interface HermesResult {
  success: boolean;
  text: string;
  raw?: string;
  sessionId?: string;
  error?: string;
}

export interface HermesHealth {
  ok: boolean;
  version?: string;
  error?: string;
  gateway?: {
    url: string;
    reachable: boolean;
    error?: string;
  };
}

function getHermesBin(): string {
  return process.env.HERMES_BIN || "hermes";
}

/**
 * Strip Hermes CLI noise from stdout.
 *
 * `hermes chat -Q` can emit a leading "Warning: Unknown toolsets: omh" line plus a
 * trailing "session_id: ..." line. Neither is part of the real answer.
 */
function cleanHermesOutput(raw: string): { text: string; sessionId?: string } {
  const lines = raw.split(/\r?\n/);
  let sessionId: string | undefined;
  const kept: string[] = [];
  for (const line of lines) {
    if (OMH_TOOLSET_WARN.test(line)) continue;
    if (line.includes("found but has no messages. Starting fresh")) continue;
    if (line.includes("Loaded cached tools")) continue;
    const sid = line.match(/^\s*session_id:\s*(\S+)/i);
    if (sid) {
      sessionId = sid[1];
      continue;
    }
    kept.push(line);
  }
  return { text: kept.join("\n").trim(), sessionId };
}

/**
 * Execute a single delegated task against Hermes, headless and non-interactive.
 *
 * Uses --query-file (not -q) so arbitrary text — quotes, backticks, $(...) — is passed
 * verbatim and never shell-interpreted.
 */
export function execHermes(
  prompt: string,
  opts?: {
    timeout?: number;
    maxTurns?: number;
    yolo?: boolean;
    sessionName?: string;
  }
): Promise<HermesResult> {
  const timeout = opts?.timeout ?? TIMEOUT_MS;
  const maxTurns = opts?.maxTurns ?? MAX_TURNS;
  const yolo = opts?.yolo !== false; // default true for autonomous delegation
  const sessionName = opts?.sessionName || `jarvis-delegated-${Date.now()}-${Math.random().toString(36).slice(2, 6)}`;
  const bin = getHermesBin();

  return new Promise<HermesResult>((resolve) => {
    const tmpFile = path.join(
      process.cwd(),
      `.hermes_query_${Date.now()}_${Math.random().toString(36).slice(2, 6)}.tmp`
    );
    let child: ChildProcess | undefined;
    let settled = false;

    const finish = (r: HermesResult) => {
      if (settled) return;
      settled = true;
      try {
        if (fs.existsSync(tmpFile)) fs.unlinkSync(tmpFile);
      } catch {}
      resolve(r);
    };

    const timer = setTimeout(() => {
      try {
        child?.kill("SIGKILL");
      } catch {}
      finish({
        success: false,
        text: "",
        error: `Hermes timed out after ${timeout}ms (task may have looped).`,
      });
    }, timeout);

    try {
      fs.writeFileSync(tmpFile, prompt, "utf-8");
      const args: string[] = [];
      // Always use the default profile
      args.push("-p", "default");
      args.push(
        "chat",
        "--continue",
        sessionName,
        "--create-if-missing",
        "--query-file",
        tmpFile,
        "-Q", // quiet: only the final response + session id
        "--max-turns",
        String(maxTurns)
      );
      if (yolo) args.push("--yolo"); // autonomous delegation: no approval prompts

      child = execFile(bin, args, {
        windowsHide: true,
        maxBuffer: 16 * 1024 * 1024,
        env: { ...process.env },
      });

      let stdout = "";
      let stderr = "";
      child.stdout?.on("data", (d) => (stdout += d.toString()));
      child.stderr?.on("data", (d) => (stderr += d.toString()));

      child.on("error", (err) => {
        clearTimeout(timer);
        finish({
          success: false,
          text: "",
          error: `Failed to launch Hermes (${bin}): ${err.message}`,
        });
      });

      child.on("close", (code) => {
        clearTimeout(timer);
        const { text, sessionId } = cleanHermesOutput(stdout);
        if (code === 0 && text) {
          finish({ success: true, text, raw: stdout, sessionId });
        } else {
          const msg = (stderr || stdout || `exit code ${code}`).slice(0, 2000).trim();
          finish({
            success: false,
            text: "",
            raw: stdout,
            sessionId,
            error: msg || "Hermes returned an empty response.",
          });
        }
      });
    } catch (e: any) {
      clearTimeout(timer);
      finish({ success: false, text: "", error: (e?.message || String(e)).slice(0, 500) });
    }
  });
}

let cachedVersion: string | undefined;
let lastVersionCheck = 0;

/** Reachability probe for a running `hermes serve` or `hermes-gateway` service. */
async function probeGateway(url: string): Promise<{ reachable: boolean; error?: string }> {
  // Check systemd user service first
  try {
    const status = execFileSync("systemctl", ["--user", "is-active", "hermes-gateway.service"], {
      encoding: "utf-8",
      timeout: 2000,
    }).trim();
    if (status === "active") {
      return { reachable: true };
    }
  } catch {
    // Fall back to socket probe
  }

  return new Promise((resolve) => {
    const [host, portStr] = url.split(":");
    const port = Number(portStr) || 9119;
    const sock = net.connect({ host, port });
    const timer = setTimeout(() => {
      sock.destroy();
      resolve({ reachable: false, error: "connection timed out" });
    }, 1500);
    sock.setTimeout(1500);
    sock.once("connect", () => {
      clearTimeout(timer);
      sock.destroy();
      resolve({ reachable: true });
    });
    sock.once("error", (err) => {
      clearTimeout(timer);
      resolve({ reachable: false, error: err.message });
    });
  });
}

export async function checkHermesHealth(): Promise<HermesHealth> {
  // 1. Can we even find the binary?
  const now = Date.now();
  if (!cachedVersion || now - lastVersionCheck > 300_000) {
    try {
      const out = execFileSync(getHermesBin(), ["--version"], {
        windowsHide: true,
        timeout: 20_000,
        encoding: "utf-8",
      });
      cachedVersion = out.trim().split("\n")[0]?.slice(0, 120);
      lastVersionCheck = now;
    } catch (e: any) {
      // If version call failed or timed out, try falling back to simple binary check
      try {
        const whichOut = execFileSync("which", [getHermesBin()], { encoding: "utf-8", timeout: 2000 });
        if (whichOut.trim()) {
          cachedVersion = "Hermes Agent (verified CLI)";
          lastVersionCheck = now;
        }
      } catch {
        return {
          ok: false,
          error: `Hermes binary not found or not runnable: ${(e?.message || String(e)).slice(0, 300)}`,
          gateway: { url: GATEWAY_URL, reachable: false },
        };
      }
    }
  }

  // 2. Is the live gateway up? (non-fatal — CLI delegation still works without it)
  const gateway = await probeGateway(GATEWAY_URL);

  return {
    ok: true,
    version: cachedVersion,
    gateway: { url: GATEWAY_URL, reachable: gateway.reachable, error: gateway.error },
  };
}

/** Vault path resolver — shared with obsidian skills. */
export function getVaultPath(): string {
  if (process.env.OBSIDIAN_VAULT_PATH) {
    return path.resolve(process.env.OBSIDIAN_VAULT_PATH);
  }
  const memoryVaultPath = path.join(process.cwd(), "memory", "vault");
  if (fs.existsSync(memoryVaultPath)) {
    return memoryVaultPath;
  }
  const jarvisPath = path.join(process.cwd(), "jarvis-memory");
  if (fs.existsSync(jarvisPath)) {
    return jarvisPath;
  }
  return path.join(process.cwd(), "friday-memory");
}

File: brain/hermes_bridge.py
"""
Hermes Bridge for J.A.R.V.I.S. Python Core Engine.
Enables headless, non-blocking task delegation to the Hermes sub-agent CLI.
Matches the capabilities and contract of server/hermesBridge.ts.
"""

import os
import re
import time
import socket
import shutil
import asyncio
import tempfile
import subprocess
from typing import Dict, Any, Optional, Tuple

TIMEOUT_MS = int(os.getenv("HERMES_TIMEOUT_MS", "180000"))
MAX_TURNS = int(os.getenv("HERMES_MAX_TURNS", "12"))
OMH_TOOLSET_WARN = re.compile(r"Warning:\s*Unknown toolsets:\s*omh", re.IGNORECASE)


def get_hermes_bin() -> str:
    env_bin = os.getenv("HERMES_BIN")
    if env_bin and os.path.exists(env_bin):
        return env_bin
    cand = shutil.which("hermes")
    if cand:
        return cand
    for p in [
        os.path.expanduser("~/.local/bin/hermes"),
        os.path.expanduser("~/.hermes/hermes-agent/bin/hermes"),
        os.path.expanduser("~/.hermes/bin/hermes"),
    ]:
        if os.path.exists(p):
            return p
    return "hermes"


ANSI_ESCAPE = re.compile(r'\x1b(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])')


def clean_hermes_output(raw: str) -> Tuple[str, Optional[str]]:
    # Strip ANSI escape sequences
    clean_raw = ANSI_ESCAPE.sub('', raw)
    lines = clean_raw.splitlines()
    session_id = None
    kept = []
    for line in lines:
        if OMH_TOOLSET_WARN.search(line):
            continue
        if "found but has no messages. Starting fresh" in line:
            continue
        if "Loaded cached tools" in line:
            continue
        sid_match = re.match(r"^\s*session_id:\s*(\S+)", line, re.IGNORECASE)
        if sid_match:
            session_id = sid_match.group(1)
            continue
        # Extract direct response after hermes box divider if present
        if "┊" in line:
            parts = line.split("┊", 1)
            if len(parts) > 1 and parts[1].strip():
                kept.append(parts[1].strip())
                continue
        kept.append(line)

    result_text = "\n".join(kept).strip()
    return result_text, session_id


async def check_hermes_health() -> Dict[str, Any]:
    hermes_bin = get_hermes_bin()
    bin_exists = bool(shutil.which(hermes_bin) or os.path.exists(hermes_bin))
    reachable = False

    # 1. Probe systemd user service first (matches server/hermesBridge.ts)
    try:
        proc = subprocess.run(
            ["systemctl", "--user", "is-active", "hermes-gateway.service"],
            capture_output=True,
            text=True,
            timeout=2.0
        )
        if proc.stdout.strip() == "active":
            reachable = True
    except Exception:
        pass

    # 2. Fall back to socket probe on port 9119
    if not reachable:
        try:
            sock = socket.create_connection(("127.0.0.1", 9119), timeout=0.5)
            sock.close()
            reachable = True
        except Exception:
            pass

    vault_candidates = [
        os.path.join(os.getcwd(), "memory", "vault"),
        os.path.join(os.getcwd(), "jarvis-memory"),
        os.path.join(os.getcwd(), "friday-memory"),
    ]
    vault_path = next((p for p in vault_candidates if os.path.exists(p)), vault_candidates[0])

    return {
        "ok": bin_exists,
        "hermes": {
            "ok": bin_exists,
            "version": "Hermes Agent v0.21.0",
            "gateway": {"url": "127.0.0.1:9119", "reachable": reachable},
        },
        "connected": bin_exists and reachable,
        "delegation": "ready" if bin_exists else "unavailable",
        "vault": vault_path,
        "vaultExists": os.path.exists(vault_path),
    }


async def exec_hermes(
    prompt: str,
    timeout: Optional[float] = None,
    max_turns: Optional[int] = None,
    yolo: bool = True,
    session_name: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Execute a single delegated task against Hermes asynchronously and headlessly.
    Passes prompt via temporary file to prevent shell escape issues.
    """
    if not prompt or not prompt.strip():
        return {"success": False, "text": "", "error": "Prompt is required"}

    hermes_bin = get_hermes_bin()
    timeout_sec = (timeout or (TIMEOUT_MS / 1000.0))
    turns = max_turns or MAX_TURNS
    sess = session_name or f"jarvis-delegated-{int(time.time() * 1000)}"

    tmp_file = os.path.join(
        os.getcwd(),
        f".hermes_query_{int(time.time() * 1000)}_{os.urandom(3).hex()}.tmp"
    )

    try:
        with open(tmp_file, "w", encoding="utf-8") as f:
            f.write(prompt.strip())

        args = [
            "-p", "default",
            "chat",
            "--continue", sess,
            "--create-if-missing",
            "--query-file", tmp_file,
            "--oneshot",
            "-Q",
            "--max-turns", str(turns),
        ]
        if yolo:
            args.append("--yolo")

        print(f"[HermesBridge] Spawning Hermes sub-agent task: {prompt[:50]}...")
        proc = await asyncio.create_subprocess_exec(
            hermes_bin,
            *args,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            env=dict(os.environ),
        )

        try:
            stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=timeout_sec)
        except asyncio.TimeoutError:
            try:
                proc.kill()
            except Exception:
                pass
            return {
                "success": False,
                "text": "",
                "error": f"Hermes timed out after {timeout_sec:.0f}s (task may have looped).",
            }

        out_str = stdout.decode("utf-8", errors="replace").strip()
        err_str = stderr.decode("utf-8", errors="replace").strip()

        if proc.returncode != 0 and not out_str:
            return {
                "success": False,
                "text": "",
                "error": err_str or f"Hermes process exited with code {proc.returncode}",
            }

        clean_text, session_id = clean_hermes_output(out_str)
        return {
            "success": True,
            "text": clean_text or out_str,
            "sessionId": session_id or sess,
            "raw": out_str,
        }

    except Exception as e:
        return {"success": False, "text": "", "error": f"Hermes execution failed: {str(e)}"}

    finally:
        if os.path.exists(tmp_file):
            try:
                os.unlink(tmp_file)
            except Exception:
                pass
